pkg/blockstorage/azure/auth.go

```diff
@@ -0,0 +1,84 @@
+package azure
+
+import (
+	"context"
+
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	"github.com/kanisterio/kanister/pkg/blockstorage"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// types of Azure creds: https://docs.microsoft.com/en-us/azure/developer/go/azure-sdk-authorization
+const (
+	CredTypeManagedIdentity   string = "cred-type-msi"
+	CredTypeClientCredentials string = "cred-type-client-credentials"
+	CredTypeClientCertificate string = "cred-type-client-certificate"
+	CredTypeUsernamePasswords string = "cred-type-username-password"
+	CredTypeDeviceToken       string = "cred-type-device-token"
+)
+
+var availableCredsType = sets.NewString(CredTypeManagedIdentity)
+
+type CredsValidity int
+
+const (
+	CredValidityUnknown CredsValidity = 0
+	CredsTypeNotSupported
+	CredsFailedAuthentication
+	CredsPassedAuthentication
+)
+
+func isCredTypeSupported(credType string) bool {
+	return availableCredsType.Has(credType)
+}
+
+func isClientCredsAvailable(config map[string]string) bool {
+	if config[blockstorage.AzureTenantID] != "" &&
+		config[blockstorage.AzureCientID] != "" &&
+		config[blockstorage.AzureClentSecret] != "" {
+		return true
+	} else {
+		return false
+	}
+}
+
+func isMSICredsAvailable(config map[string]string) bool {
+	if config[blockstorage.AzureTenantID] == "" &&
+		config[blockstorage.AzureCientID] != "" &&
+		config[blockstorage.AzureClentSecret] == "" {
+		return true
+	} else {
+		return false
+	}
+}
+
+// internal interface to authenticate with different Azure credentials type
+type authenticator interface {
+	authenticate(creds map[string]string) (CredsValidity, error)
+}
+
+// authenticate with MSI creds
+type msiAuthenticator struct{}
+
+func (m *msiAuthenticator) authenticate(creds map[string]string) (CredsValidity, error) {
```

Comment:
This interface is not united tested because it involves network calls which will results in `context timeout`